baseapp/baseapp_test.go

```diff
@@ -2153,3 +2153,122 @@ func TestBaseApp_Init(t *testing.T) {
 		require.Equal(t, tc.expectedSnapshot.KeepRecent, tc.bapp.snapshotManager.GetKeepRecent())
 	}
 }
+
+func TestGenerateFraudProof(t *testing.T) {
+
+	/*
+		1. Create a fresh baseapp with a tracekv store
+		2. Create a 'block' and put transactions that set certain key/value pairs in it
+		3. We should be able to `generate fraud proof` IF block.hash not the same, trigger fraud:
+			- go through all transactions, run them, and keep track of which substores are being used
+			- export the SMTs inside those substores into a fraud proof data structure along with block height
+
+
+		Tests to write:
+
+		1. Block with bad txs: Txs that exceed gas limits, validateBasic fails, unregistered messages (see TestRunInvalidTransaction)
+		2. Block with invalid appHash at the end
+		3. Corrupted Fraud Proof: bad SMT format, insufficient key-value pairs inside SMT needed to verify fraud
+		4. Bad block, fraud proof needed, fraud proof works, chain halts
+
+		TODO:
+		Figure out how the tracekv interacts with the SMT multistore - done
+		Try to keep tx as generic as possible so you don't need to care about the messages inside a Tx
+
+
+		Question: What is a block? Right now abstract, need to make it more concrete
+		Candidate Answer: For now, Block is just a list of a transactions and a header with an app hash
+
+		Note: In the current implementation, all substores might not be SMTs, do we assume they are?
+
+	*/
+
+	codec := codec.NewLegacyAmino()
+	registerTestCodec(codec)
+
+	storeTraceBuf := &bytes.Buffer{}
+	subStoreTraceBuf := &bytes.Buffer{}
+
+	routerOpt := func(bapp *BaseApp) {
+		bapp.Router().AddRoute(sdk.NewRoute(routeMsgKeyValue, func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
+			kv := msg.(*msgKeyValue)
+			bapp.cms.GetKVStore(capKey2).Set(kv.Key, kv.Value)
+			return &sdk.Result{}, nil
+		}))
+	}
+
+	app := setupBaseApp(t,
+		AppOptionFunc(routerOpt),
+		SetSubstoreTracer(storeTraceBuf),
+		SetTracerFor(capKey2, subStoreTraceBuf),
```

Comment:
If we trace which keys where modified "globally" per app, won't we always get too many key/value pairs? I.e. shouldn't the scope of a trace be per block instead of "globally" per app? Or do we reset this regularly?

Reply:
So this is part of generating a fraudProof, ideally in the real world, when a new baseApp is setup up by an optimint full node when it is gossiped a fraudProof by the P2P network, then ONLY we will do any sort of tracing and we'd have the keys you refer to inside the fraudProof which is minimal

Reply:
Thanks for the explanation! Makes sense.